Add Arteco ZS-304Z soil moisture sensor

ehbello opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->

Adds support for the Arteco ZS-304Z (TS0601) soil moisture sensor (_TZE284_o9ofysmo).

This device provides:
- Soil moisture (DP 3)
- Temperature (DP 5)
- Humidity (DP 101)
- Illuminance (DP 102)
- Battery state (DP 14, mapped to percentage remaining)

It also includes various configuration parameters:
- Soil sampling interval (DP 103)
- Calibration for soil, humidity, illuminance, and temperature (DPs 104-107)
- Soil warning threshold (DP 110)
- Water warning binary sensor (DP 111)


## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->

<img width="1019" height="881" alt="image" src="https://github.com/user-attachments/assets/899f0917-891f-4dc1-a0cc-f68c4c564e8f" />

- Fixes #4541

## Device diagnostics
<!--
  Diagnostics data is used by ZHA unit tests to make sure quirks create the expected
  entities and we do not have regressions. For all new quirks, we require device
  diagnostics.

  You can find the diagnostics information by going to the device page, clicking the
  three dots, and then by clicking on "Download diagnostics". Drag-and-drop the
  downloaded file into this section.
-->

[zha-01KCVF5Y1M7D1DM1GHG4QQPBW9-_TZE284_o9ofysmo TS0601-dab25e8a31cf620cc69fcf8ffc1d34ce.json](https://github.com/user-attachments/files/24361061/zha-01KCVF5Y1M7D1DM1GHG4QQPBW9-_TZE284_o9ofysmo.TS0601-dab25e8a31cf620cc69fcf8ffc1d34ce.json)

## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works
- [x] Device diagnostics data has been attached

ehbello committed:
feat(tuya): Add Arteco ZS-304Z soil moisture sensor

Adds support for the Arteco ZS-304Z (TS0601) soil moisture sensor
(_TZE284_o9ofysmo).

This device provides:
- Soil moisture (DP 3)
- Temperature (DP 5)
- Humidity (DP 101)
- Illuminance (DP 102)
- Battery state (DP 14, mapped to percentage remaining)

It also includes various configuration parameters:
- Soil sampling interval (DP 103)
- Calibration for soil, humidity, illuminance, and temperature
  (DPs 104-107)
- Soil warning threshold (DP 110)
- Water warning binary sensor (DP 111)

Signed-off-by: Enrique Hernández Bello <ehbello@gmail.com>

ehbello committed:
test(tuya): Add data handling test for Arteco ZS-304Z soil sensor

This test validates the `handle_get_data` method of the TuyaMCUCluster
for the Arteco ZS-304Z soil sensor (model `_TZE284_o9ofysmo`).
It ensures correct parsing and attribute mapping for several Tuya DPs:
- DP 3: Soil Moisture to `soil_moisture.measured_value`
- DP 5: Temperature to `temperature.measured_value`
- DP 14: Battery to `power.battery_percentage_remaining`
- DP 101: Humidity to `humidity.measured_value`
- DP 102: Illuminance to `illuminance.measured_value`

Signed-off-by: Enrique Hernández Bello <ehbello@gmail.com>

ehbello committed:
feat(tuya-sensor): Improve Arteco ZS-304Z soil moisture sensor definition

Refactors the Arteco ZS-304Z (TS0601) soil moisture sensor to use the
generic tuya_sensor builder method. This change explicitly defines the
sensor's type, attribute name, unit, and device class, enabling ZHA to
correctly interpret the raw DP 3 value (e.g., 50 for 50%) as a
percentage moisture sensor.

Previously, a proprietary tuya_soil_moisture method was used, which
internally scaled the value. Now, the raw DP value is exposed via the
tuya_manufacturer cluster, aligning with standard ZHA entity creation
for better integration. The associated test has been updated to assert
against the raw value directly in the tuya_manufacturer cluster.

Signed-off-by: Enrique Hernández Bello <ehbello@gmail.com>

tegkhanna commented:
Hey folks, can we merge this if there are no blockers. If there are, could you state those please ?

hcoohb commented:
Would this be the same as #4599  ?

Also the battery seems to be handled differently

tegkhanna commented:
AFAIU, these 2 are similar and based on the same z2mqtt entity. AFAIK, the battery is better on this one. Tuya device have an enum representation but not exact battery value. The author would know better

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -333,3 +333,90 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
 )
+
+(
+    TuyaQuirkBuilder("_TZE284_o9ofysmo", "TS0601")  # Arteco ZS-304Z
```
```suggestion
    TuyaQuirkBuilder("_TZE284_o9ofysmo", "TS0601")  # Arteco ZS-304Z
    .applies_to("Arteco", "ZS-304Z")
```
otherwise it does not apply to that device, right?

ehbello replied:
It makes sense. I did not specify that variant because mine is detected as TS0601 and I cannot test others. I just applied your suggestion.

SuperSandro2000 replied:
I think it might depend on if you have the z2m or the pure tuya variant.

ehbello replied:
I purchased the z2m variant.

ehbello replied:
According to [this](https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/tuya.ts#L9729) zigbee2mqtt implementation, this fingerprint `_TZE284_o9ofysmo` actually corresponds to an **Arteco ZS-301Z** :thinking:

ehbello committed:
feat(tuya): add Arteco ZS-304Z soil sensor variant

Add the "Arteco" "ZS-304Z" model and manufacturer pair to the existing
soil sensor quirk. Update the test suite to verify functionality for
both model identifiers using parametrization.

Signed-off-by: Enrique Hernández Bello <ehbello@gmail.com>

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        attribute_name="soil_calibration",
+        type=t.int16s,
+        entity_type=EntityType.CONFIG,
+        translation_key="soil_calibration",
+        fallback_name="Soil calibration",
+    )
+    .tuya_number(
+        dp_id=105,
+        attribute_name="humidity_calibration",
+        type=t.int16s,
+        entity_type=EntityType.CONFIG,
+        translation_key="humidity_calibration",
+        fallback_name="Humidity calibration",
+    )
+    .tuya_number(
+        dp_id=106,
+        attribute_name="illuminance_calibration",
+        type=t.int16s,
+        entity_type=EntityType.CONFIG,
+        translation_key="illuminance_calibration",
+        fallback_name="Illuminance calibration",
+    )
+    .tuya_number(
+        dp_id=107,
+        attribute_name="temperature_calibration",
+        type=t.int16s,
+        multiplier=0.1,
+        step=0.1,
+        entity_type=EntityType.CONFIG,
+        translation_key="temperature_calibration",
+        fallback_name="Temperature calibration",
+    )
+    .tuya_number(
+        dp_id=110,
+        attribute_name="soil_warning",
+        type=t.uint16_t,
+        entity_type=EntityType.CONFIG,
+        translation_key="soil_warning",
+        fallback_name="Soil warning threshold",
+    )
```
Do all of these miss unit, min_value, max_value and step?

Based on https://github.com/zigpy/zha-device-handlers/pull/4599 I would add this:

ehbello replied:
I just applied your suggestions regarding those parameters to all of them.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+    .tuya_number(
+        dp_id=104,
+        attribute_name="soil_calibration",
+        type=t.int16s,
```
```suggestion
        type=t.int16s,
        unit=PERCENTAGE,
        min_value=-30,
        max_value=30,
        step=1,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+    .tuya_number(
+        dp_id=105,
+        attribute_name="humidity_calibration",
+        type=t.int16s,
```
```suggestion
        type=t.int16s,
        unit=PERCENTAGE,
        min_value=-30,
        max_value=30,
        step=1,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+    .tuya_number(
+        dp_id=103,
+        attribute_name="soil_sampling",
+        type=t.uint16_t,
```
```suggestion
        type=t.uint16_t,
        unit=UnitOfTime.SECONDS,
        min_value=5,
        max_value=360,
        step=1,
```

not sure if we should allow values like 600? Is that device dependent?

ehbello replied:
According to [this](https://www.zigbee2mqtt.io/devices/ZS-304Z.html#soil-sampling-numeric) the zigbee2mqtt implementation allows up to 3600 as the maximum allowed.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+    .tuya_number(
+        dp_id=106,
+        attribute_name="illuminance_calibration",
+        type=t.int16s,
```
```suggestion
        type=t.int16s,
        unit=LIGHT_LUX,
        min_value=-1000,
        max_value=1000,
        step=1,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        dp_id=107,
+        attribute_name="temperature_calibration",
+        type=t.int16s,
+        multiplier=0.1,
```
```suggestion
        multiplier=0.1,
        min_value=-2,
        max_value=2,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+    .tuya_number(
+        dp_id=110,
+        attribute_name="soil_warning",
+        type=t.uint16_t,
```
```suggestion
        type=t.uint16_t,
        unit=PERCENTAGE,
        min_value=0,
        max_value=100,
        step=1,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        dp_id=14,
+        ep_attribute=TuyaPowerConfigurationCluster2AAA.ep_attribute,
+        attribute_name="battery_percentage_remaining",
+        converter=lambda x: {0: 20, 1: 100, 2: 200}.get(x, 200),
```
Not sure if we should fall back to full or low battery here.

ehbello replied:
Maybe "Middle"? like... something "undefined" between _empty_ and _full_

SuperSandro2000 commented:
Before my suggestions HASS showed:

<img width="163" height="384" alt="image" src="https://github.com/user-attachments/assets/fd7da5f3-0a74-4692-ab51-3eba7966c103" />


and after:

<img width="336" height="787" alt="image" src="https://github.com/user-attachments/assets/87f416f5-ed36-4e1b-85d3-c47e44d6865f" />


The calibration entities are none functional for me, I cannot change the slider.

<img width="587" height="612" alt="image" src="https://github.com/user-attachments/assets/2ea864c0-276e-4aa5-87cd-ab9fd329f110" />

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+    )
+    .tuya_number(
+        dp_id=104,
+        attribute_name="soil_calibration",
```
Should this be soil_moisture_calibration or is that missing?

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        type=t.int16s,
+        multiplier=0.1,
+        step=0.1,
+        entity_type=EntityType.CONFIG,
```
```suggestion
        entity_type=EntityType.CONFIG,
        device_class=NumberDeviceClass.TEMPERATURE,
```

ehbello replied:
I used `NumberDeviceClass.TEMPERATURE_DELTA`  here (rather than `NumberDeviceClass.TEMPERATURE)` because this DP is **temperature calibration** — an **offset** applied to the reported measurement, not the ambient temperature reading itself.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        dp_id=106,
+        attribute_name="illuminance_calibration",
+        type=t.int16s,
+        entity_type=EntityType.CONFIG,
```
```suggestion
        entity_type=EntityType.CONFIG,
        device_class=NumberDeviceClass.ILLUMINANCE,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        dp_id=105,
+        attribute_name="humidity_calibration",
+        type=t.int16s,
+        entity_type=EntityType.CONFIG,
```
```suggestion
        entity_type=EntityType.CONFIG,
        device_class=NumberDeviceClass.HUMIDITY,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        dp_id=104,
+        attribute_name="soil_calibration",
+        type=t.int16s,
+        entity_type=EntityType.CONFIG,
```
if https://github.com/zigpy/zha-device-handlers/pull/4602/files#r3041411838 is true

```suggestion
        entity_type=EntityType.CONFIG,
        device_class=NumberDeviceClass.MOISTURE,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        dp_id=103,
+        attribute_name="soil_sampling",
+        type=t.uint16_t,
+        entity_type=EntityType.CONFIG,
```
```suggestion
        entity_type=EntityType.CONFIG,
        device_class=NumberDeviceClass.DURATION,
```

ehbello replied:
Done.

SuperSandro2000 commented on zhaquirks/tuya/tuya_sensor.py:
```diff
@@ -372,3 +372,91 @@ class ServerCommandDefs(TuyaMCUCluster.ServerCommandDefs):
+        dp_id=110,
+        attribute_name="soil_warning",
+        type=t.uint16_t,
+        entity_type=EntityType.CONFIG,
```
This is probably wrong:

```suggestion
        entity_type=EntityType.CONFIG,
        device_class=NumberDeviceClass.MOISTURE,
```

ehbello replied:
Done.

SuperSandro2000 commented:
Strangely the calibration settings fixed themselves after deploying the device classes and waiting a bit. I was just trying to change the type to int8s and before deploying that it somehow started to work.

ehbello committed:
feat(tuya): improve MCU sensor configuration and calibration entities

Update Tuya MCU sensor configuration entities with proper units,
range limits, and device classes. This includes:
- Renaming soil_calibration to soil_moisture_calibration.
- Adding duration, moisture, humidity, illuminance, and temperature
  delta classes.
- Defining min/max values and steps for all calibration entities.

Co-authored-by: SuperSandro2000 <sandro.jaeckel@gmail.com>
Signed-off-by: Enrique Hernández Bello <ehbello@gmail.com>
